File: android/app/src/main/java/com/novawave/synth/audio/Envelope.kt
package com.novawave.synth.audio

enum class EnvStage { IDLE, ATTACK, DECAY, SUSTAIN, RELEASE }

/** Classic ADSR envelope generator, linear attack/release with exponential-feel decay. */
class Envelope(private val sampleRate: Float) {

    var stage: EnvStage = EnvStage.IDLE
        private set
    private var level: Float = 0f

    var attackSec: Float = 0.005f
    var decaySec: Float = 0.15f
    var sustainLevel: Float = 0.7f
    var releaseSec: Float = 0.2f

    fun noteOn() {
        stage = EnvStage.ATTACK
    }

    fun noteOff() {
        if (stage != EnvStage.IDLE) stage = EnvStage.RELEASE
    }

    val isActive: Boolean get() = stage != EnvStage.IDLE

    fun nextSample(): Float {
        when (stage) {
            EnvStage.IDLE -> level = 0f
            EnvStage.ATTACK -> {
                val step = 1f / (attackSec.coerceAtLeast(0.0005f) * sampleRate)
                level += step
                if (level >= 1f) {
                    level = 1f
                    stage = EnvStage.DECAY
                }
            }
            EnvStage.DECAY -> {
                val step = (1f - sustainLevel) / (decaySec.coerceAtLeast(0.0005f) * sampleRate)
                level -= step
                if (level <= sustainLevel) {
                    level = sustainLevel
                    stage = EnvStage.SUSTAIN
                }
            }
            EnvStage.SUSTAIN -> level = sustainLevel
            EnvStage.RELEASE -> {
                val step = level / (releaseSec.coerceAtLeast(0.0005f) * sampleRate)
                level -= step
                if (level <= 0.0005f) {
                    level = 0f
                    stage = EnvStage.IDLE
                }
            }
        }
        return level
    }
}

File: src/audio/Envelope.ts
export type EnvStage = 'idle' | 'attack' | 'decay' | 'sustain' | 'release';

/** ADSR envelope advanced by real elapsed time (control-rate, ticked from rAF). */
export class Envelope {
  stage: EnvStage = 'idle';
  private level = 0;

  attackSec = 0.005;
  decaySec = 0.15;
  sustainLevel = 0.7;
  releaseSec = 0.2;

  get value(): number {
    return this.level;
  }

  get isActive(): boolean {
    return this.stage !== 'idle';
  }

  noteOn(): void {
    this.stage = 'attack';
  }

  noteOff(): void {
    if (this.stage !== 'idle') this.stage = 'release';
  }

  tick(dt: number): number {
    switch (this.stage) {
      case 'idle':
        this.level = 0;
        break;
      case 'attack': {
        const step = dt / Math.max(this.attackSec, 0.0005);
        this.level += step;
        if (this.level >= 1) {
          this.level = 1;
          this.stage = 'decay';
        }
        break;
      }
      case 'decay': {
        const step = ((1 - this.sustainLevel) * dt) / Math.max(this.decaySec, 0.0005);
        this.level -= step;
        if (this.level <= this.sustainLevel) {
          this.level = this.sustainLevel;
          this.stage = 'sustain';
        }
        break;
      }
      case 'sustain':
        this.level = this.sustainLevel;
        break;
      case 'release': {
        const step = (this.level * dt) / Math.max(this.releaseSec, 0.0005);
        this.level -= step;
        if (this.level <= 0.0005) {
          this.level = 0;
          this.stage = 'idle';
        }
        break;
      }
    }
    return this.level;
  }
}
